clang/lib/AST/StmtPrinter.cpp

```diff
@@ -1257,28 +1257,48 @@ void StmtPrinter::VisitConstantExpr(ConstantExpr *Node) {
 }
 
 void StmtPrinter::VisitDeclRefExpr(DeclRefExpr *Node) {
-  if (const auto *OCED = dyn_cast<OMPCapturedExprDecl>(Node->getDecl())) {
+  ValueDecl *VD = Node->getDecl();
+  if (const auto *OCED = dyn_cast<OMPCapturedExprDecl>(VD)) {
     OCED->getInit()->IgnoreImpCasts()->printPretty(OS, nullptr, Policy);
     return;
   }
-  if (const auto *TPOD = dyn_cast<TemplateParamObjectDecl>(Node->getDecl())) {
+  if (const auto *TPOD = dyn_cast<TemplateParamObjectDecl>(VD)) {
     TPOD->printAsExpr(OS, Policy);
     return;
   }
   if (NestedNameSpecifier *Qualifier = Node->getQualifier())
     Qualifier->print(OS, Policy);
   if (Node->hasTemplateKeyword())
     OS << "template ";
-  if (Policy.CleanUglifiedParameters &&
-      isa<ParmVarDecl, NonTypeTemplateParmDecl>(Node->getDecl()) &&
-      Node->getDecl()->getIdentifier())
-    OS << Node->getDecl()->getIdentifier()->deuglifiedName();
-  else
-    Node->getNameInfo().printName(OS, Policy);
+  DeclarationNameInfo NameInfo = Node->getNameInfo();
+  if (IdentifierInfo *ID = NameInfo.getName().getAsIdentifierInfo();
+      ID || NameInfo.getName().getNameKind() != DeclarationName::Identifier) {
+    if (Policy.CleanUglifiedParameters &&
+        isa<ParmVarDecl, NonTypeTemplateParmDecl>(VD) && ID)
+      OS << ID->deuglifiedName();
+    else
+      NameInfo.printName(OS, Policy);
+  } else {
+    switch (VD->getKind()) {
+    case Decl::NonTypeTemplateParm: {
+      auto *TD = cast<NonTypeTemplateParmDecl>(VD);
+      OS << "value-parameter-" << TD->getDepth() << '-' << TD->getIndex() << "";
+      break;
+    }
+    case Decl::ParmVar: {
+      auto *PD = cast<ParmVarDecl>(VD);
+      OS << "function-parameter-" << PD->getFunctionScopeDepth() << '-'
+         << PD->getFunctionScopeIndex();
+      break;
+    }
+    default:
+      llvm_unreachable("Unhandled anonymous declaration kind");
```

Comment:
FYI I have some downstream code where this is reached with VD->getKind() == Decomposition. I don't understand the problem well yet so I can't tell if this is a problem with our code or with this change. Commenting here in case this sounds like a case that should be handled here.

Reply:
This could probably point to some problem in your code.

1) It's strange that you have a DeclRefExpr pointing to a decomposition, instead of one of its bindings.
2) I'd like to understand your use case. We can't refer to a decomposition in source code, so there is no sensible thing the StmtPrinter can do. Before this change, we printed an empty string for any anonymous entity, which was not helpful.

There is likely to be some test case in your downstream where an empty string is being printed and that's not good either.

Is that something which could be reproduced upstream?

We could add some workaround support for printing these, something along the lines of `decomposition-<name-of-first-binding>`.

Reply:
Our test is trying to print the control flow graph of a translation unit that looks like this:
```
#include <vector>
#include <utility>

int main() {
  std::vector<std::pair<int, int>> v;
  for (auto& [a, b]: v) {
    a;
    b;
  }
}
```
Building that file with `clang -Xclang -analyze -Xclang -analyzer-checker=debug.DumpCFG -fsyntax-only` triggers the same crash.

Reply:
Yeah this is similar to the other issue here: https://github.com/llvm/llvm-project/pull/124605#issuecomment-2620280655

On current clang, this test case prints a couple of steps which are completely empty.

Reply:
Sure, but current clang doesn't crash. In our case, we had a test that ensures that something compiles and as a byproduct dumped a lot of information to a log. A couple of empty lines in the log are not ideal, but a crash is a breaking change.

Reply:
Fixed here: https://github.com/llvm/llvm-project/pull/125001